WORKSPACE

```diff
@@ -128,10 +136,86 @@ load("@bazel_features//:deps.bzl", "bazel_features_deps")
 
 bazel_features_deps()
 
+load("@toolchains_llvm//toolchain:deps.bzl", "bazel_toolchain_dependencies")
+
+bazel_toolchain_dependencies()
+
 load("@build_bazel_apple_support//lib:repositories.bzl", "apple_support_dependencies")
 
 apple_support_dependencies()
 
+load("@toolchains_llvm//toolchain:rules.bzl", "llvm_toolchain")
+
+COMMON_EXTRA_COMPILE_FLAGS = [
+    "-Wextra",
+    "-Werror",
+    "-pedantic",
+    "-Wno-unused-parameter",
+    "-fPIC",
+]
+
+LINUX_EXTRA_COMPILE_FLAGS = COMMON_EXTRA_COMPILE_FLAGS + ["-gz=zlib"]
+
+MACOS_EXTRA_COMPILE_FLAGS = COMMON_EXTRA_COMPILE_FLAGS + [
+    "-Wno-shorten-64-to-32",
+    "-Wno-gcc-compat",
+    "-Wno-missing-field-initializers",
+    "-Wno-unused-private-field",
+]
+
+COMMON_EXTRA_CXX_FLAGS = [
+    "-Wformat=2",
+    "-Wno-unused-parameter",
+    "-pthread",
+    "-Wno-deprecated-enum-enum-conversion",
+]
+
+LINUX_EXTRA_CXX_FLAGS = COMMON_EXTRA_CXX_FLAGS + ["-Wno-psabi"]
+
+MACOS_EXTRA_CXX_FLAGS = COMMON_EXTRA_CXX_FLAGS + [
+    "-Wno-unused-const-variable",
+    "-Wno-error=c11-extensions",
+    "-Wno-deprecated-anon-enum-enum-conversion",
+]
+
+LINUX_EXTRA_LINK_FLAGS = ["-Wl,-rpath,$ORIGIN"]
+
+MACOS_EXTRA_LINK_FLAGS = ["-Wl,-rpath,@loader_path"]
+
+llvm_toolchain(
+    name = "llvm_toolchain",
+    cxx_standard = {"": "c++20"},
+    extra_compile_flags = {
+        "linux-aarch64": LINUX_EXTRA_COMPILE_FLAGS,
+        "linux-x86_64": LINUX_EXTRA_COMPILE_FLAGS,
+        "macos-aarch64": MACOS_EXTRA_COMPILE_FLAGS,
+        "macos-x86_64": MACOS_EXTRA_COMPILE_FLAGS,
+    },
+    extra_cxx_flags = {
+        "linux-aarch64": LINUX_EXTRA_CXX_FLAGS,
+        "linux-x86_64": LINUX_EXTRA_CXX_FLAGS,
+        "macos-aarch64": MACOS_EXTRA_CXX_FLAGS,
+        "macos-x86_64": MACOS_EXTRA_CXX_FLAGS,
+    },
+    extra_link_flags = {
+        "linux-aarch64": LINUX_EXTRA_LINK_FLAGS,
+        "linux-x86_64": LINUX_EXTRA_LINK_FLAGS,
+        "macos-aarch64": MACOS_EXTRA_LINK_FLAGS,
+        "macos-x86_64": MACOS_EXTRA_LINK_FLAGS,
+    },
+    llvm_version = "19.1.7",
```

Comment:
No particular reason for picking LLVM 19 specifically. It's just easier to specify a version that `toolchains_llvm` [already knows about](https://github.com/bazel-contrib/toolchains_llvm/blob/master/toolchain/internal/llvm_distributions.bzl#L32), and I was initially experimenting with toolchains_llvm 1.4.0 due to a bug in 1.5.0 with RBE.

Reply:
I would recommend we go with 20.1 or newer due to https://github.com/llvm/llvm-project/issues/110231

Reply:
Ooh yikes. Can do, since we're pulling in an unreleased toolchains_llvm anyway.

Reply:
Wait no, the clang 18 build won't work on Ubuntu 24.04.
```
external/llvm_toolchain_llvm/bin/clang: error while loading shared libraries: libtinfo.so.5: cannot open shared object file: No such file or directory
```